library/modules/EventManager.cpp

```diff
@@ -703,293 +913,426 @@ static void manageBuildingEvent(color_ostream& out) {
+                item0.wound_id == item1.wound_id)
+                continue;
+            // add to list because item is not equipped in the same way
+            equipmentChanges.emplace(tick, InventoryChangeData(unit->id, new InventoryItem(item_old.itemId, item_old.item), new InventoryItem(itemId, *ct_item)));
+        }
+        //check for dropped items
+        for (auto &key_value : last_tick_inventory) {
+            auto &item = key_value.second;
+            // add to list if unit doesn't have the item
+            if (currentlyEquipped.find(item.itemId) == currentlyEquipped.end()) {
+                equipmentChanges.emplace(tick, InventoryChangeData(unit->id, new InventoryItem(item.itemId, item.item), nullptr));
+            }
+        }
+
+        //update equipment
+        last_tick_inventory.clear();
+        for (auto ct_item : current_tick_inventory) {
+            auto itemId = ct_item->item->id;
+            last_tick_inventory.emplace(itemId, InventoryItem(itemId, *ct_item));
+        }
+    }
+    // iterate event handler callbacks
+    multimap<Plugin*,EventHandler> copy(handlers[EventType::INVENTORY_CHANGE].begin(), handlers[EventType::INVENTORY_CHANGE].end());
+    for (auto &key_value : copy) {
+        auto &handler = key_value.second;
+        auto last_tick = eventLastTick[handler];
+        // ensure handler's callback frequency
+        if (tick - last_tick >= handler.freq) {
+            eventLastTick[handler] = tick;
+            // send all new equipment changes since it last fired
+            for(auto iter = equipmentChanges.upper_bound(last_tick); iter != equipmentChanges.end(); ++iter){
+                handler.eventHandler(out, &iter->second);
+            }
+        }
+    }
+}
+
+static void manageReportEvent(color_ostream& out) {
+    if (!df::global::world)
+        return;
+
+    int32_t tick = df::global::world->frame_counter;
+    std::unordered_set<int32_t> valid_reports;
+    // update reports list
+    for(auto &r : df::global::world->status.reports){
+        valid_reports.emplace(r->id);
+        newReports.emplace(tick, r->id);
+    }
+
+    // iterate event handler callbacks
+    multimap<Plugin*,EventHandler> copy(handlers[EventType::REPORT].begin(), handlers[EventType::REPORT].end());
+    for (auto &key_value : copy) {
+        auto &handler = key_value.second;
+        auto last_tick = eventLastTick[handler];
+        // enforce handler's callback frequency
+        if (tick - last_tick >= handler.freq) {
+            eventLastTick[handler] = tick;
+            // send all new reports since it last fired
+            auto iter = newReports.upper_bound(last_tick);
+            for(;iter != newReports.end();) {
+                if (valid_reports.count(iter->second)) {
+                    handler.eventHandler(out, (void*) intptr_t(iter->second));
+                    ++iter;
+                } else {
+                    iter = newReports.erase(iter->second);
+                }
```

Comment:
It's a bit unclear how useful this actually is. Users would still need to find the report from the id. So I think this is a bit redundant. Worst case scenario, we've taken away useful information. Best case we've prevented a user from erroneously assuming the id exists